feat: add `stats/array/max-by` (#7130)
stdlib-js/stdlib: lib/node_modules/@stdlib/stats/array/max-by/README.md

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# maxBy

> Calculate the maximum value of an array via a callback function.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var maxBy = require( '@stdlib/stats/array/max-by' );
```

#### maxBy( x, clbk\[, thisArg] )

Computes the maximum value of an array via a callback function.

```javascript
function accessor( v ) {
    return v * 2.0;
}

var x = [ -2.0, 1.0, 3.0, -5.0, 4.0, 0.0, -1.0, -3.0 ];

var v = maxBy( x, accessor );
// returns 8.0
```

The function has the following parameters:

-   **x**: input array.
-   **clbk**: callback function.
-   **thisArg**: execution context (_optional_).

> The invoked callback is provided four arguments:
> 
> -   **value**: array element.
> -   **aidx**: array index.

Reviewer: ```suggestion
The invoked callback is provided three arguments:

-   **value**: current array element.
-   **index**: current array index.
```

> -   **array**: input array/collection.

Reviewer: ```suggestion
-   **array**: input array.
```

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# maxBy

> Calculate the maximum value of an array via a callback function.

<section class="intro">

</section>

<!-- /.intro -->

<section class="usage">

## Usage

```javascript
var maxBy = require( '@stdlib/stats/array/max-by' );
```

#### maxBy( x, clbk\[, thisArg] )

Computes the maximum value of an array via a callback function.

```javascript
function accessor( v ) {
    return v * 2.0;
}

var x = [ -2.0, 1.0, 3.0, -5.0, 4.0, 0.0, -1.0, -3.0 ];

var v = maxBy( x, accessor );
// returns 8.0
```

The function has the following parameters:

-   **x**: input array.
-   **clbk**: callback function.
-   **thisArg**: execution context (_optional_).

The invoked callback is provided four arguments:

-   **value**: array element.
-   **aidx**: array index.
-   **sidx**: strided index (`offset + aidx*stride`).

> -   **array**: input array/collection.

Reviewer: This isn't correct. But also, I see why you did this. I've updated the implementation to demonstrate how we address.

To set the callback execution context, provide a `thisArg`.

```javascript
function accessor( v ) {
    this.count += 1;
    return v * 2.0;
}

var x = [ -2.0, 1.0, 3.0, -5.0, 4.0, 0.0, -1.0, -3.0 ];

var context = {
    'count': 0
};

var v = maxBy( x, accessor, context );
// returns 8.0

var cnt = context.count;
// returns 8
```

</section>

<!-- /.usage -->

<section class="notes">

## Notes

-   If provided an empty array, the function returns `NaN`.
-   A provided callback function should return a numeric value.
-   If a provided callback function does not return any value (or equivalently, explicitly returns `undefined`), the value is **ignored**.
-   The function supports array-like objects having getter and setter accessors for array element access (e.g., [`@stdlib/array/base/accessor`][@stdlib/array/base/accessor]).

> -   When possible, prefer using [`dmax`][@stdlib/stats/array/dmax], [`smax`][@stdlib/stats/array/smax], and/or [`max`][@stdlib/stats/array/max], as, depending on the environment, these interfaces are likely to be significantly more performant.

Author: The packages `dmax` & `smax` are yet to be added.
Should they be mentioned here?
I think we can keep it as link attached would be same after addition of package.

> -   When possible, prefer using [`dmax`][@stdlib/stats/array/dmax], [`smax`][@stdlib/stats/array/smax], and/or [`max`][@stdlib/stats/array/max], as, depending on the environment, these interfaces are likely to be significantly more performant.

Reviewer: ```suggestion
```

We won't be adding those packages, so this comment should be removed in its entirety.

Reviewer: The `stats/array/*` namespace is intentionally only "generic".
